### ISSUE TITLE
Exploration task for using Chisme in the charm

### ISSUE THREAD

**Issue by mvlassis** [canonical/github-profiles-automator#13]
### Context

This exploration task follows #6, and includes exploration of incorporating the helpers of [Chisme](https://github.com/canonical/charmed-kubeflow-chisme) in the final charm.``

### What needs to get done

1. Look into how chisme operates, and what parts of chisme will be used in the final part
2.  Specify how the final charm will be broken into chisme components

### Definition of Done

We can specify in the final spec what components the final charm will consist of

**Comment by mvlassis** [comment]
## Configuration
The final charm will have these 4 configuration options:
- `repository`: Holds the repository of the Github repo to be cloned
- `period`: How long to wait between syncs in seconds
- `pmr-yaml-path`: The relative path to the `.yaml` file that holds the PMR information
- `ssh-key-secret-id`: This will hold the ID for the `ssh-key` Juju secret. This will be handled similar to `tls-secret-id` in [istio-pilot](https://github.com/canonical/istio-operators/blob/8fa311e98f7e366378abc93d63c31974d4c003a6/charms/istio-pilot/src/charm.py#L127) 

## Components
2 components will be created in the  `src/components` directory. One named `GitSyncPebbleServicet` will handle the Pebble service, and one named `PMRHandlerComponent` will create the PMR from the profile information and call its handlers.

## GitSyncPebbleService (PebbleServiceComponent)
This component is responsible for creating the `git-sync` Pebble service, it needs the following 3 inputs which will live in a [Python dataclass](https://docs.python.org/3/library/dataclasses.html):
- REPOSITORY: This will hold the value of `self.config["repository"]`
- PERIOD: This will hold the value of `self.config["period"]
- SSH_KEY: This will receive the value of the Juju secret using `self.model.get_secret[self.config["ssh-key-secret-id"]`

The component will then push the SSH_KEY to the repository, probably using a [ContainerFileTemplate](https://github.com/canonical/charmed-kubeflow-chisme/blob/b64f6ccca228d08c6533cbe83382306cd46670dc/src/charmed_kubeflow_chisme/components/pebble_component.py#L117), and configure the Pebble layer with the following command:
```
        command = ' '.join(
            [
                'sleep 1.1 && ', # See https://github.com/canonical/pebble/issues/240
                '/git-sync',
                f"--repo={inputs.REPOSITORY}",
                '--depth=1',
                f"--period={inputs.PERIOD}s",
                '--link=cloned-repo',
                '--root=/git',
                '--ssh-known-hosts=false',
                '--verbose=9',
                '--exechook-command=/wrapper-script.sh'
            ]
        )
```
Note that the `sleep 1.1` command is a way to circumvent a [known issue in Pebble](https://github.com/canonical/pebble/issues/240) where Pebble doesn't restart a service if it immediately exits. 

##  PMRHandlerComponent
This component will use the mounted storage named `content-from-git`, It will use as input a dataclass named `PMRInputs` with the following fields:
- PMR_YAML_PATH: This will hold the value of `self.config["pmr_yaml_path"]`
This component will check the validity of the `yaml` file in the path, and set the appropriate status if the `.yaml` is invalid.
We have to define the `events_to_observe` method on this class to  watch the [pebble_custom_notice](https://ops.readthedocs.io/en/latest/state-transition-testing.html#ops.testing.CharmEvents.pebble_custom_notice) event in addition to the standard events that `Charmreconciler` watches for.

## Alternatives Considered
 I thought of creating a separate `Component` named `SSHKeyComponent` that would handle the SSH key Juju secret. This component would be responsible for receiving the `ssh-key-secret-id` configuration option and pushing it to the workload container. @DnPlas suggested that we can do this in the `PebbleServiceComponent` by adding it to the `files_to_push` option.

Thanks to @DnPlas and @kimwnasptd for providing useful feedback.

**Comment by kimwnasptd** [comment]
Some small nits, but overall LGTM!

1. IMO I'd rephrase `period` to `sync-period` to make it slightly more clear. 
2. Regarding `sleep 1.1`, since we'll go with JuJu 3.6.1 can't we skip this now?
3. For the `PMRHandlerComponent`, shouldn't it only be run **only** with pebble custom notices to ensure a YAML exists?

Lastly, we'll need to do some work for the Charm to be able to get the principals from KFP and Istio Gateway
https://docs.google.com/document/d/1WlRJTOs-J1ghZVygO3AOFvbETLyt_fY2h_PsafPMLa0/edit?tab=t.0#heading=h.8htmgjvabbte

**Comment by mvlassis** [comment]
Thanks @kimwnasptd:
1. `sync-period` is clearer so I have updated it
2. You're right about `sleep`, I have removed the command in both the charm code and the component
3. The `Component` pattern calls the `configure_charm` function on all components, regardless of the event that has been observed. If we want to *only* watch for this event, we will probably not wrap this logic in a component, and just observe it with:
```
framework.observe(self.on["git_sync"].pebble_custom_notice, self._on_pebble_custom_notice)
```

Thanks for noticing the relations, I will make sure to add them in the specification

**Comment by mvlassis** [comment]
After syncing with @kimwnasptd, we found that Chisme currently has the following limitations that block using them for this charm:

## PebbleServiceComponent
The `git-sync` executable may fail with an exit code of 1 if the remote repository doesn't exist, the SSH key is wrong etc. In that case, Pebble's `replan()` raises a [ChangeError](https://ops.readthedocs.io/en/latest/pebble.html#ops.pebble.ChangeError). 

We want to catch this error, and set an appropriate error message depending on the logs that Pebble produces. We also want to set the Charm's status according to the error. The issue is that:
- We cannot catch `ChangeError` *and* update the charm's status based on checking this error.

### More details
- The `_update_layer` method of [PebbleServiceComponent](https://github.com/canonical/charmed-kubeflow-chisme/blob/4a0be92ceb2d00346fa34ae87ca8744cb19e56c4/src/charmed_kubeflow_chisme/components/pebble_component.py) *doesn't catch* the ChangeError
- The `update_layer` function in [charmed_kubeflow_chisme/pebble](https://github.com/canonical/charmed-kubeflow-chisme/blob/4a0be92ceb2d00346fa34ae87ca8744cb19e56c4/src/charmed_kubeflow_chisme/pebble/_update_layer.py) *does* catch the `ChangeError`, so we cannot use that either

## PMRHandler
- This component should be responsible for updating the PMR whenever a Pebble Custom Notice is received. This means that ideally, we want its reconcile function to *only* call this logic for this event, and not for the [default events](https://github.com/canonical/charmed-kubeflow-chisme/blob/4a0be92ceb2d00346fa34ae87ca8744cb19e56c4/src/charmed_kubeflow_chisme/components/charm_reconciler.py#L108) that chisme installs.

**Comment by mvlassis** [comment]
After syncing with @DnPlas, we can do the following:

## PebbleServiceComponent
We can go around this issue by:
- Adding checks in the charms `init()` method to the check the config so that the repository given is valid
- Add health checks to check whether we can connect to Github using the given SSH key, and observe the events `PebbleCheckFailedEvent` and `PebbleCheckRecoveredEvent` to change the status of the charm depending on their status.

## PMRHandler
Do not use chisme here since we only want to call the function on specific occasions.

**Comment by kimwnasptd** [comment]
A small note we discussed is that with the above mechanism:
1. The charm won't handle the `ChangeError`, which means once `replan` happens the charm will expose the service as active
2. Once a health-check happens, or an `update-status` then the state will be checked and the charm will go to Blocked, Waiting etc

The above is an issue caused by the fact that we don't catch the `ChangeError`. But we'll live with this for now, in order to use the base-charm from Chisme into this code-base
